lib/asg-runner-stack.ts

```diff
@@ -0,0 +1,173 @@
+import * as cdk from 'aws-cdk-lib';
+import * as ec2 from 'aws-cdk-lib/aws-ec2';
+import * as iam from 'aws-cdk-lib/aws-iam';
+import * as resourcegroups from 'aws-cdk-lib/aws-resourcegroups';
+import { aws_autoscaling as autoscaling } from 'aws-cdk-lib'; 
+import { Construct } from 'constructs';
+import { RunnerType } from '../config/runner-config'; 
+import { readFileSync } from 'fs';
+import { ENVIRONMENT_STAGE } from './finch-pipeline-app-stage';
+
+interface ASGRunnerStackProps extends cdk.StackProps {
+  env: cdk.Environment | undefined;
+  stage: ENVIRONMENT_STAGE;
+  licenseArn: string;
+  type: RunnerType;
+}
+
+/**
+ * A stack to provision an autoscaling group for macOS instances. This requires:
+ *  - a self-managed license (manually created as cdk/cfn does not support this)
+ *  - a resource group
+ *  - a launch template
+ *  - an auto scaling group
+ */
+export class ASGRunnerStack extends cdk.Stack {
+  constructor(scope: Construct, id: string, props: ASGRunnerStackProps) {
+    super(scope, id, props);
+    const arch = props.type.arch === 'arm' ? 'arm64_mac' : 'x86_64_mac';
+    const instanceType = props.type.arch === 'arm' ? 'mac2.metal' : 'mac1.metal'
+        
+    if (props.env == undefined) {
+      throw new Error("Runner environment is undefined!")
+    }
+    
+    const vpc = cdk.aws_ec2.Vpc.fromLookup(this, 'VPC',{isDefault: true});
+
+    const securityGroup = new ec2.SecurityGroup(this, 'MacEC2SecurityGroup', {
+      vpc,
+      description: 'Allow SSH (TCP port 22) in',
+      allowAllOutbound: true
+    });
+    securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(22), 'Allow SSH Access');
+
+    const role = new iam.Role(this, 'MacEC2Role', {
+      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com')
+    });
+
+    role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'));
+    role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('ResourceGroupsandTagEditorFullAccess'));
+
+    // Grant EC2 instances access to secretsmanager to retrieve the GitHub api key to register runners
+    if (props.stage != ENVIRONMENT_STAGE.Beta) {
+      role.addToPolicy(new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        actions: [
+          'secretsmanager:GetResourcePolicy',
+          'secretsmanager:GetSecretValue',
+          'secretsmanager:DescribeSecret',
+          'secretsmanager:ListSecretVersionIds'
+        ],
+        resources: [ `arn:aws:secretsmanager:${props.env?.region}:${props.env?.account}:secret:${props.type.repo}-runner-reg-key*` ]
+      }))
+    }
+
+    // Create a custom name for this as names for resource groups cannot be repeated
+    const resourceGroup = new resourcegroups.CfnGroup(this, 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup', {
+      name: 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup',
```

Comment:
This is what I was suggesting, not sure if Monirul was mentioning the same thing:

```suggestion

    // Create a custom name for this as names for resource groups cannot be repeated
    const resourceGroupName = `Mac-${props.type.macOSVersion.split('.')[0]}-${arch.split('_')[0]}HostGroup`
    const resourceGroup = new resourcegroups.CfnGroup(this, resourceGroupName, {
      name: resourceGroupName,
```

Reply:
Oh! Got it, okay that makes a lot of sense. Will update, thanks!